src/Functions/punycode.cpp

```diff
@@ -0,0 +1,199 @@
+#include "config.h"
+
+#if USE_IDNA
+
+#include <Functions/FunctionFactory.h>
+#include <Functions/FunctionStringToString.h>
+
+#ifdef __clang__
+#    pragma clang diagnostic push
+#    pragma clang diagnostic ignored "-Wnewline-eof"
+#endif
+#    include <ada/idna/punycode.h>
+#    include <ada/idna/unicode_transcoding.h>
+#ifdef __clang__
+#    pragma clang diagnostic pop
+#endif
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int BAD_ARGUMENTS;
+    extern const int ILLEGAL_COLUMN;
+}
+
+struct PunycodeEncodeImpl
+{
+    static void vector(
+        const ColumnString::Chars & data,
+        const ColumnString::Offsets & offsets,
+        ColumnString::Chars & res_data,
+        ColumnString::Offsets & res_offsets)
+    {
+        const size_t rows = offsets.size();
+        res_data.resize(rows * 64); /// just a guess
+        res_offsets.resize(rows);
+
+        size_t prev_offset = 0;
+        size_t prev_res_offset = 0;
+        size_t res_data_bytes_written = 0;
+        std::u32string value_utf32;
+        std::string value_puny;
+        for (size_t row = 0; row < rows; ++row)
+        {
+            const char * value = reinterpret_cast<const char *>(&data[prev_offset]);
+            const size_t value_length = offsets[row] - prev_offset - 1;
+
+            size_t value_utf32_length = ada::idna::utf32_length_from_utf8(value, value_length);
+            value_utf32.resize(value_utf32_length, '\0');
+
+            ada::idna::utf8_to_utf32(value, value_length, value_utf32.data());
+
+            bool ok = ada::idna::utf32_to_punycode(value_utf32, value_puny);
+            if (!ok)
+                throw Exception(ErrorCodes::BAD_ARGUMENTS, "Internal error during Punycode encoding");
+
+            const size_t bytes_to_write = value_puny.size() + 1;
+            if (res_data_bytes_written + bytes_to_write > res_data.size())
+            {
+                size_t new_size = std::max(res_data.size() * 2, res_data_bytes_written + bytes_to_write);
+                res_data.resize(new_size);
+            }
+
+            std::memcpy(&res_data[res_data_bytes_written], value_puny.data(), value_puny.size());
+            res_data_bytes_written += value_puny.size();
+
+            res_data[res_data_bytes_written] = '\0';
+            res_data_bytes_written += 1;
+
+            res_offsets[row] = prev_res_offset + bytes_to_write;
+
+            prev_offset = offsets[row];
+            prev_res_offset = res_offsets[row];
+            value_utf32.clear();
+            value_puny.clear();
```

Comment:
It is not really obvious that `ada::idna::utf32_to_punycode()` and the other function requires the output parameter to be empty. I think a comment here would be useful.